Make union/except/intersect chainable as SelectManager

jsanders opened the pull request:
Fixes #118.

This is the same as #239, but against master instead of 4-0-stable. It would be really nice to get this in so we don't need to keep freshening it! Or if there's a reason it isn't being merged, feedback would be great.

ofavre committed:
Make union/except/intersect chainable as SelectManager

Fixes rails/arel#118
Fixes rails/arel#98

Conflicts:
	lib/arel/nodes/and.rb
	lib/arel/nodes/select_statement.rb
	lib/arel/visitors/mssql.rb
	lib/arel/visitors/to_sql.rb

Conflicts:
	lib/arel/nodes/and.rb
	lib/arel/nodes/binary.rb
	lib/arel/predications.rb
	lib/arel/select_manager.rb
	lib/arel/visitors/depth_first.rb
	lib/arel/visitors/mssql.rb
	lib/arel/visitors/to_sql.rb
	test/visitors/test_mysql.rb
	test/visitors/test_oracle.rb
	test/visitors/test_to_sql.rb

jsanders committed:
Remove left and right from Nary

It doesn't make sense and any uses are almost certainly buggy.

jsanders committed:
Don't allow adding children to Nary

jsanders committed:
Fix a warning about redefinition of test_Arel_Nodes_Or

amalagaura commented:
It is frustrating AREL doesn't support this, there are a lot of use cases for chaining union operators. It would be great if this was resolved.

jsanders commented:
@amalagaura: Yeah, I rely on this behavior and keep needing to update this branch to work with new versions. If the maintainers have no interest in merging this functionality, perhaps I should instead figure out a way to do it with a monkey-patch or a separate library.

@rafaelfranca / @tenderlove / other maintainers: I'm not sure if you've seen this, but a thumbs-up or thumbs-down on this functionality would be really nice, so that I have a sense for whether or not I'm wasting my time on it. If the functionality is desired, but you don't like the implementation, I'm willing to put together a different PR that speaks to that.

amalagaura commented:
This seems like a no-brainer. I also have 3 Arel::SelectManagers that I need to merge and I am forced to do a .to_sql  and join(' UNION '). 

Arel needs union to not return a Node::Binary and instead return a proper SelectManager, otherwise union really is broken since SQL definitely supports chainable unions.

sockmonk commented:
I'd also like to see a better way to work with unions. Feedback on @jsanders  pull request please?

ClayShentrup commented:
How has this been ignored a year? @tenderlove ??

tenderlove commented:
I sincerely apologize for not having time for this.  Can you please rebase it and I will review.  I'm really sorry.

jsanders commented:
@tenderlove: No worries! I'll find some time to rebase and ping you again.

ryanswood commented:
@jsanders Did you rebase? We'd love to have this. Thanks!

jsanders committed:
Fix oracle12 visitor for Except

Except is an nary node, so the visitor must use `inject_join` rather than
`infix_value`.

jsanders committed:
Update gemspec

jsanders commented:
@tenderlove: Rebased this onto current master and things look pretty good. Feedback welcome!

Sorry for the delay, folks.

tamird commented on lib/arel/select_manager.rb:
```diff
@@ -6,11 +6,15 @@ class SelectManager < Arel::TreeManager
-      @ctx    = @ast.cores.last
-      from table
+      if cores
+        @ast = Nodes::SelectStatement.new cores
```
:( guess this was ignored

jsanders replied:
Actually, I did it before, but ended up skipping it when I rebased when it caused a conflict and I decided it was pretty trivial. I guess if people feel strongly, I can add it back.

ryanswood commented:
@tenderlove Waiting on your tender PR love. Thanks.

amalagaura commented:
@tenderlove Any updates? I have another use case and don't want to use `join(' OR ')`

nilcolor commented:
Hello. What about this?
```ruby
Arel::Nodes::Union.new(select_manager_1, Arel::Nodes::Union.new(select_manager_2, select_manager_3))
```
This gives us "chained" unions. Some over-parenthesized but workable solution. We can generalize it a'lil:
```ruby
def build_union(left, right)
  if right.length > 1
    Arel::Nodes::UnionAll.new(left, build_union(right[0], right[1..-1]))
  else
    Arel::Nodes::UnionAll.new(left, right[0])
  end
end

managers = [select_manager_1, select_manager_2, select_manager_3]
build_union(managers[0], managers[1..-1]).to_sql
# => ( (SELECT table1.* from table1)
#    UNION ALL
#    ( (SELECT table2.* from table2)
#    UNION ALL
#    (SELECT table3.* from table3) ) )
```
This works in Arel 4.0.2... No need to `join` lots of `to_sql`...

matthewd commented:
Per #523, Arel development is moving to [rails/rails](https://github.com/rails/rails).

If this PR is still relevant, please consider reopening it over there.
